Ref types (#240)
oras-project/oras-www: docs/concepts/reftypes.mdx

## Implementing Reference Types in OCI artifacts

The concept of linking or relating OCI artifacts is achieved using the concept of reference types, where one artifact can be set as a subject to another. In this context, "subject" refers to the target of a reference.

### The Art of Associating Artifacts

Consider a typical cloud-native environment with countless OCI artifacts. Often, it's imperative to manage relationships between them—say, an OCI artifact's signatures, or metadata about an artifact. This is where reference types come into the picture.

A reference type enables an OCI artifact to refer to another, creating a defined link or association between them. This association can be for various purposes, such as versioning, updates, signatures, or dependencies.

![OCI artifact with refererrs pointing to the artifact](/img/concepts/reftypes.png)

## Creating a Reference Artifact

OCI Artifacts may be stored in a content-addressable storage system or OCI Layout directory. For the purpose of this guide, we'll use the OCI Distribution Specification to create and list reference types.

The creation of a reference artifact is a two-step process:

> 1. **Step One**: Put an artifact, say Artifact A, using the PUT Manifest and blob upload APIs if the artifact has blob content.

Reviewer: nit: a bulleted list might be better than a numbered list since the step number is explicitly stated
```suggestion
- **Step One**: Put an artifact, say Artifact A, using the PUT Manifest and blob upload APIs if the artifact has blob content.
```

The list may be paginated if the number of referrers is large.

```mermaid
sequenceDiagram
    participant C as Client
    participant R as Registry
    C->>R: GET /v2/<name>/referrers/<digest>
    R-->>C: 200 OK (List of all referrers)
```

### Listing referrers with a specific artifact type

In this sequence, the client requests a list of referrers to a specific artifact of a certain media type from the registry, which then responds with the filtered list.
The response will include the `OCI-Filters-Applied: artifactType` header to indicate that the response is filtered by the `artifactType` query parameter.

The artifact type is determied by the same rules as per [OCI artifacts guidlines](artifact.mdx#determining-the-artifact-type).

```mermaid
sequenceDiagram
    participant C as Client
    participant R as Registry
    C->>R: GET /v2/<name>/referrers/<digest>?artifactType=<artifactType>
    R-->>C: 200 OK (List of referrers of specific media type)
```

The request below lists shows an example for listing referrers of a specific media type.

```json

> GET /v2/hello-world/referrers/sha256:5e140a6..?artifactType=application/example HTTP/1.1

Reviewer: nit: `application/example` should be URL encoded as `application%2Fexample`

// response
Content-Type: application/vnd.oci.image.index.v1+json
Oci-Filters-Applied: artifactType

{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:a10d2cf503458e48996a3e3030c75fb5f5cd967e21a38ca69ce6c1e1dee5fd27",
      "size": 677,
      "annotations": {
        "org.opencontainers.image.created": "2023-08-05T07:08:43Z"
      },
      "artifactType": "application/example"
    }
  ]
}
```

## Fetching the reference artifact

Once the referrers have been queried, the client may fetch the reference artifact using the digest that is returned in the list of referrers.
The client may then use the digest to fetch the artifact contents using the [pull workflow][pull-workflow] as described in the distribution specification.

```mermaid
sequenceDiagram
    participant C as Client
    participant R as Registry

    Note over C,R: Pull Workflow Starts

>     C->>R: GET /v2/<name>/manifests/<tag>

Reviewer: Is it `digest` instead of `tag`?
```suggestion
    C->>R: GET /v2/<name>/manifests/<digest>
```

    R-->>C: 200 OK, return manifest

    Note over C,R: If manifest has config

    C->>R: GET /v2/<name>/blobs/<digest> (Config)

>     R-->>C: 200 OK, return config

Reviewer: It might not be `200 OK` for blob downloads. It could be `307 Redirect`.

    Note over C,R: If manifest has blobs

    C->>R: GET /v2/<name>/blobs/<digest> (Blob 1)
    R-->>C: 200 OK, return blob 1

    C->>R: GET /v2/<name>/blobs/<digest> (Blob 2)
    R-->>C: 200 OK, return blob 2

    Note over C,R: Save to OCI Layout or file system

    C->>+C: Save manifest, config and blobs 
    Note over C: Pull Workflow Ends
```

## The Power of Associations

The ability to create and list referrers opens a world of possibilities in managing and tracking OCI Artifacts.
From versioning to updates, and from dependencies to signatures, the use of reference types ensures that we can effectively maintain the complex web of relationships between various OCI Artifacts in a cloud-native environment.

> To further explore this concept, dive deeper into the [OCI Distribution Specification](https://github.com/opencontainers/distribution-spec/blob/main/spec.md) and the [OCI Image Specification](https://github.com/opencontainers/image-spec).

Reviewer: The links to the `image-spec` and to the `distribution-spec` are not consistent.

These comprehensive guides will provide more insights into the use of reference types and other details of managing OCI Artifacts.

[listing-referrers]: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#listing-referrers
[pull-workflow]: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pull
[push-workflow]: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#push

> [tag-schema]: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#referrers-tag-schema

Reviewer: A tagged version is better than the `main` branch for the URLs.

---
title: Attached Artifacts
sidebar_position: 1
---

# Attaching and listing reference types

> The OCI distribution specification allows us to "link" [OCI artifacts](artifact.mdx)

Reviewer 2: Should we state the specific version of the OCI Distribution Spec? I prefer `
Since the [OCI Distribution Specification v1.1.0-rc.1](https://github.com/opencontainers/distribution-spec/releases/tag/v1.1.0-rc1), it allows us to "link" [OCI artifacts](artifact.mdx)`

Author: I'll add the details to the latter part in a subsequent section. Pointing to a point in time concept and liking to the spec in the first sentence seems too dense.

Author: Also the concept of distribution is not needed to create refereces since its can be on the client side also without using references.

through the use of the concept of reference types.
Artifacts can be tied to one another by defining one as a subject of another.
This is crucial for maintaining dependencies between different artifacts and
enabling the movement of artifacts and their dependencies between environments.

## Implementing Reference Types in OCI artifacts

The concept of linking or relating OCI artifacts is achieved using the concept
of reference types, where one artifact can be set as a subject to another.
In this context, "subject" refers to the target of a reference.

### The Art of Associating Artifacts

Consider a typical cloud-native environment with countless OCI artifacts.
Often, it's imperative to manage relationships between them—say, an OCI artifact's
signatures, or metadata about an artifact. This is where reference types come into the picture.

A reference type enables an OCI artifact to refer to another, creating a defined
link or association between them. This association can be for various purposes,

> such as versioning, updates, signatures, or dependencies.

Reviewer 2: What does `versioning` and `updates` mean? Could you clarify it?

Can we revise it to:

A reference type enables an OCI artifact to refer to another, creating a defined
link or association between them. This association can be for various purposes,
such as versioning, supply chain artifacts (signatures, SBOM, etc.), and dependencies.

![OCI artifact with refererrs pointing to the artifact](/img/concepts/reftypes.png)

## Creating a Reference Artifact

OCI Artifacts may be stored in a content-addressable storage system or OCI Layout directory.
For the purpose of this guide, we'll use the OCI Distribution Specification to
create and list reference types.

The creation of a reference artifact is a two-step process:

1. **Step One**: Put an artifact, say Artifact A, using the PUT Manifest and
blob upload APIs if the artifact has blob content.

>     - The prerequisite for a reference type is that the subject artifact must exist in the repository.

Reviewer 2: ```suggestion
> [!NOTE]  
> The prerequisite for a reference type is that the subject artifact must exist in the repository.
```

Author: Actually now that I read it back, there is no requirement that the subject should exist :) I need to remove that. Thanks for the catch.

Author: Added a note to highlight that its not required at the bottom.

2. **Step Two**: Put another artifact, say Artifact B, which refers to Artifact A.
In this case, Artifact A is the 'subject' of Artifact B.

>     - This is done using the [push workflow][push-workflow] in the distribution specification.

Reviewer 2: > [!NOTE]  
> This is done using the [push workflow][push-workflow] in the distribution specification.

>     - This is done using the [push workflow][push-workflow] in the distribution specification.

Reviewer 2: ```suggestion
> [!NOTE]  
> This is done using the [push workflow][push-workflow] in the distribution specification.This is done using the [push workflow][push-workflow] in the distribution specification.
```

  "config": {
    "mediaType": "application/vnd.oci.empty.v1+json",
    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
    "size": 2
  },
  "layers": [
    {
      "mediaType": "application/vnd.oci.image.layer.v1.tar",
      "digest": "sha256:d2a84f4b8b650937ec8f73cd8be2c74add5a911ba64df27458ed8229da804a26",
      "size": 12,
      "annotations": {
        "org.opencontainers.image.title": "attach/hello.txt"
      }
    }
  ],
  "subject": {
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "digest": "sha256:5e140a61e16155b30356685a6801e5250339bfb11370e70573d28d4ff2dc89cf",
    "size": 477
  },
  "annotations": {
    "org.opencontainers.image.created": "2023-08-05T06:58:38Z"
  }
}
```

## Listing Referrers

Once an artifact has one or more referring artifacts, it's often necessary to
list these referrers for various purposes such as tracking dependencies, updates
or obtaining metadata of the artifact.
The OCI [Distribution Specification][listing-referrers] defines an API endpoint
(`/v2/<name>/referrers/<digest>`) for this purpose.

If the registry does not support listing referrers then the [referrers tag schema][tag-schema]

> may be used to obtain and static list Index of refererres.

Reviewer 2: ```suggestion
may be used to obtain and statically list the Index of referrers. 
```

Author: update this to - 
If the registry does not support listing referrers, then the [referrers tag schema][tag-schema]
may be used to obtain a client created non-dynamic index of refererrers.

}
```

## Fetching the reference artifact

Once the referrers have been queried, the client may fetch the reference artifact
using the digest that is returned in the list of referrers.
The client may then use the digest to fetch the artifact contents using the
[pull workflow][pull-workflow] as described in the distribution specification.

```mermaid
sequenceDiagram
    participant C as Client
    participant R as Registry

    Note over C,R: Pull Workflow Starts

    C->>R: GET /v2/<name>/manifests/<digest>
    R-->>C: 200 OK, return manifest

    Note over C,R: If manifest has config

    C->>R: GET /v2/<name>/blobs/<digest> (Config)
    R-->>C: 200 OK, return config

    Note over C,R: If manifest has blobs

    C->>R: GET /v2/<name>/blobs/<digest> (Blob 1)
    R-->>C: 200 OK/307 Redirect, return blob 1

    C->>R: GET /v2/<name>/blobs/<digest> (Blob 2)
    R-->>C: 200 OK/307 Redirect return blob 2

    Note over C,R: Save to OCI Layout or file system

    C->>+C: Save manifest, config and blobs 
    Note over C: Pull Workflow Ends
```

## The Power of Associations

The ability to create and list referrers opens a world of possibilities in
managing and tracking OCI Artifacts.

> From versioning to updates, and from dependencies to signatures, the use of

Reviewer 2: The design of Reference Types (Association) also empowers the following scenarios: 

- Supply chain artifacts discovery and distribution
- Artifact portability, users can easily copy a graph of OCI content across registries
- Maintain artifact lifecycle graph. For example, delete the root artifact, the associated signatures, SBOMs can all be deleted as well

      },
      "artifactType": "application/example"
    }
  ]
}
```

## Fetching the reference artifact

Once the referrers have been queried, the client may fetch the reference artifact
using the digest that is returned in the list of referrers.
The client may then use the digest to fetch the artifact contents using the
[pull workflow][pull-workflow] as described in the distribution specification.

```mermaid
sequenceDiagram
    participant C as Client
    participant R as Registry

    Note over C,R: Pull Workflow Starts

    C->>R: GET /v2/<name>/manifests/<digest>
    R-->>C: 200 OK, return manifest

    Note over C,R: If manifest has config

    C->>R: GET /v2/<name>/blobs/<digest> (Config)
    R-->>C: 200 OK, return config

    Note over C,R: If manifest has blobs

    C->>R: GET /v2/<name>/blobs/<digest> (Blob 1)
    R-->>C: 200 OK/307 Redirect, return blob 1

    C->>R: GET /v2/<name>/blobs/<digest> (Blob 2)
    R-->>C: 200 OK/307 Redirect return blob 2

    Note over C,R: Save to OCI Layout or file system

    C->>+C: Save manifest, config and blobs 
    Note over C: Pull Workflow Ends
```

## The Power of Associations

> The introduction of Reference Types(Associations) opens a world of possibilities in

Reviewer: nit on spacing
```suggestion
The introduction of Reference Types (Associations) opens a world of possibilities in
```

managing and linking of OCI Artifacts while also empowering the following scenarios: 

- Discovery and distribution of artifacts like SBOMs or signatures for supply chain. 
- Movement of a graph of OCI content across environments. 

> - Management of the life cycle of a graph of aritifacts enabling retaining or deletion

Reviewer: What is `xx graph`?

Reviewer: We might want to reword this sentence so that readers can understand more easily.

Reviewer: Lifecycle management of a graph of artifacts (enabling retaining or deleting a sub-graph of content?). I'm confused.

Author: Let me know if the reworking helps?  @shizhMSFT

## Implementing Reference Types in OCI artifacts

The concept of linking or relating OCI artifacts is achieved using the concept
of reference types, where one artifact can be set as a subject to another.
In this context, "subject" refers to the target of a reference.

### The Art of Associating Artifacts

Consider a typical cloud-native environment with countless OCI artifacts.
Often, it's imperative to manage relationships between them—say, an OCI artifact's
signatures, or metadata about an artifact. This is where reference types come into the picture.

A reference type enables an OCI artifact to refer to another, creating a defined
link or association between them. This association can be for various purposes,
such as versioning, updates, signatures, or dependencies.

![OCI artifact with refererrs pointing to the artifact](/img/concepts/reftypes.png)

## Creating a Reference Artifact

OCI Artifacts may be stored in a content-addressable storage system or OCI Layout directory.
For the purpose of this guide, we'll use the OCI Distribution Specification to
create and list reference types.

The creation of a reference artifact is a two-step process:

> 1. **Step One**: Put an artifact, say Artifact A, using the PUT Manifest and

Reviewer: nit: the previous comment was marked resolved but no change happened.
```suggestion
- **Step One**: Put an artifact, say Artifact A, using the PUT Manifest and
```

blob upload APIs if the artifact has blob content.
    - The prerequisite for a reference type is that the subject artifact must exist in the repository.

2. **Step Two**: Put another artifact, say Artifact B, which refers to Artifact A.
In this case, Artifact A is the 'subject' of Artifact B.

> > [!NOTE]  

Reviewer: The `> [!NOTE]` mark only works in GitHub rendered pages. Apparently, the docusaurus does not render it properly.

<img width="716" alt="image" src="https://github.com/oras-project/oras-www/assets/32161882/16943597-ab65-4d6b-91ee-b207388610c7">

Author: I've updated it to use docusarus style :::tip but also moved that content out of the note and instead made a call out the fact that subject does not need to exist.
